Add state_events option for state module functions

barneysowood opened the pull request:
### What does this PR do?

Adds a state_events option to state.highstate, state.apply, state.sls, state.sls_id to enable state events whilst applying states.

Removes the need to set state_events globally in the master config, without requiring adding fire_event to every state.

The ability to send state events is really useful but under-utilised. I suspect this is because you either have to enable it globally in the master config 

This is a pre-requisite to some work I'm doing to improve feedback in the salt cli command

### Previous Behavior
State events could only be enabled globally in the master config

### New Behavior
State events can be enabled on a per highstate or state.apply basis.

### Merge requirements satisfied?

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
No

barneysowood committed:
Add state_events option for state module functions

Adds a state_events option to state.highstate, state.apply, state.sls,
state.sls_id to enable state events whilst applying states.

Removes the need to set state_events globally in the master config,
without requiring adding fire_event to every state.

barneysowood committed:
Add tests for state_events flag in state module

nicholasmhughes commented:
re-run pr-macosx-catalina-x86_64-py3-pytest

nicholasmhughes commented on salt/state.py:
```diff
@@ -3742,7 +3742,8 @@ def __gen_opts(self, opts):
             opts["env_order"] = mopts.get("env_order", opts.get("env_order", []))
             opts["default_top"] = mopts.get("default_top", opts.get("default_top"))
-            opts["state_events"] = mopts.get("state_events")
+            if "state_events" not in opts:
```
**nitpick (non-blocking):** this could be an `or` like the line below

barneysowood replied:
Good point - updated that so it's consistent.

garethgreenaway commented:
> re-run pr-macosx-catalina-x86_64-py3-pytest

MacOS test failures are a known failure right now due to a licensing issue that we're in the process of resolving.

barneysowood committed:
Update state_events opt handling to be consistent

Updates the state_events opts handling to be consistent with how other
opts are processed in state.py

barneysowood committed:
Add changelog entry

barneysowood committed:
Add documentation on the state_event option

Adds additional documentation on the state_events option to the
Salt Master Events doc

barneysowood committed:
Adds documentation on fire_events to master events

Adds documentation on the fire_events state option to the
Salt Master Events doc for completeness.

barneysowood commented:
@nicholasmhughes and @whytewolf - I've added docs and a changelog entry for this - if you could review when you get a moment, that would be great.

nicholasmhughes commented on tests/pytests/integration/modules/state/test_state_state_events.py:
```diff
@@ -0,0 +1,107 @@
+log = logging.getLogger(__name__)
+
+
+EVENT_TAG = "salt/job/*/prog/{}/0"
```
**chore (blocking):** globals are discouraged in pytests. please convert to a fixture.

garethgreenaway replied:
@barneysowood Just a heads up about the comment from @nicholasmhughes

barneysowood replied:
@nicholasmhughes replaced that with a fixture - let me know if that looks OK.

nicholasmhughes commented on salt/modules/state.py:
```diff
@@ -705,6 +705,12 @@ def apply_(mods=None, **kwargs):
+        The state_events option sends progress events as each function in
+        a state run completes execution.
+
+        .. versionadded:: 3006.0
```
**question (blocking):** this one is for @Ch3LL , @garethgreenaway , @whytewolf , or anyone else that knows the state of the 3006.0 freeze... can this one sneak in?

garethgreenaway replied:
@nicholasmhughes We should be able to get this in for 3006.

barneysowood replied:
Great :)

barneysowood committed:
Change state_event tag from global to fixture
